tokio-util/src/sync/cancellation_token.rs

```diff
@@ -222,3 +254,85 @@ impl<'a> Future for WaitForCancellationFuture<'a> {
         }
     }
 }
+
+// ===== impl WaitForCancellationFutureOwned =====
+
+impl core::fmt::Debug for WaitForCancellationFutureOwned {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.debug_struct("WaitForCancellationFutureOwned").finish()
+    }
+}
+
+impl Future for WaitForCancellationFutureOwned {
+    type Output = ();
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<()> {
+        let this = unsafe { Pin::into_inner_unchecked(self) };
+
+        loop {
+            if this.cancellation_token.is_cancelled() {
+                return Poll::Ready(());
+            }
+
+            // No wakeups can be lost here because there is always a call to
+            // `is_cancelled` between the creation of the future and the call to
+            // `poll`, and the code that sets the cancelled flag does so before
+            // waking the `Notified`.
+            if this
+                .get_future_mut()
+                // Safety:
+                //
+                // `self` is pinned, so self.future is also pinned.
+                .map(|fut| unsafe { Pin::new_unchecked(fut) })
+                .map(|fut| fut.poll(cx).is_pending())
+                .unwrap_or_default()
+            {
+                return Poll::Pending;
+            }
+
+            let notified = this.cancellation_token.inner.notified();
+
+            // Safety:
+            //  - We transmute notified into Notified<'static>
+            //    since there's no 'static
+            //  - self is pinned, so it's safe to have self-reference
+            //    data structure.
+            //  - this.cancellation_token contains an Arc to TreeNode,
+            //    which is guaranteed to have stable dereference.
+            //    So even if `Notified` implements `Unpin` and `self`
+            //    get moved, it would still be valid.
+            this.future = Some(unsafe { mem::transmute(notified) });
+        }
+    }
+}
+
+impl WaitForCancellationFutureOwned {
+    fn do_drop<'a>(&'a mut self) {
+        if let Some(future) = self.future.take() {
+            // Safety:
+            //
+            // The future itself refererences cancellation_token, so its
+            // lifetime must be at least as long as 'a.
+            let future: tokio::sync::futures::Notified<'a> = unsafe { mem::transmute(future) };
+            drop(future);
+        }
+    }
```

Comment:
This moves the `Notified` future, but it is pinned so it may not be moved.

Reply:
Thanks for pointing out!
I've fixed with `ManuallyDrop` so that the future is dropped in place.